src/util/bip341.rs

```diff
@@ -0,0 +1,556 @@
+// Rust Bitcoin Library
+// Written in 2021 by
+//   The rust-bitcoin developers
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! BIP341 Implementation
+//!
+//! Implementation of [Bip341](https://github.com/bitcoin/bips/blob/150ab6f5c3aca9da05fccc5b435e9667853407f4/bip-0341.mediawiki)
+//! signature message
+//!
+
+use consensus::Encodable;
+use hashes::{sha256, Hash};
+use std::{fmt, io};
+use util::taproot::{TapLeafHash, TapSighashHash};
+use {Script, SigHashType, Transaction, TxOut};
+
+/// Contains outputs of previous transactions to provide to the [SigHashCache::signature_hash]
+/// method, In the case [SigHashType] variant is `ANYONECANPAY`, [Prevouts::Anyone] may be provided
+pub enum Prevouts {
+    /// When modifier `ANYONECANPAY` is provided, only prevout of the current input is needed
+    Anyone(usize, TxOut),
```

Comment:
nit: Add a doccomment about `usize` being the current index of input

Reply:
done in 5e4b64c

Reply:
would it make sense to represent as a u32 as that matches Outpoint better?